File: packages/scene-wgpu/src/wgpuDebugPrelude.ts
import { getWgpuRenderStateRuntime } from '@flighthq/render-wgpu';
import type { WgpuRenderState } from '@flighthq/types';

import type { WgpuMeshPipeline } from './wgpuMeshPipeline';
import {
  createWgpuMeshPipeline,
  ensureWgpuPlaceholderTextureView,
  ensureWgpuScenePipeline,
  WGPU_MESH_PRELUDE_WGSL,
} from './wgpuMeshPipeline';
import type { WgpuMaterialBinding } from './wgpuSceneRuntime';
import { getWgpuSceneRuntime } from './wgpuSceneRuntime';

// The shared Wgpu debug prelude — the WGSL mirror of scene-gl's glDebugPrelude. One module source for
// every lighting-INDEPENDENT debug/utility pass material (Depth, Normal). WGSL has no preprocessor, so
// a `MODE` const discriminator (DEPTH_MODE / NORMAL_MODE) and a `HAS_NORMAL_MAP` const the pipeline
// compiler folds select the fragment branch and the (future) normal-map path. Both modes output LINEAR
// color with no lighting term, matching the rgba16float scene target, and ignore the Frame light block.
//
// Depth mode reads the perspective-encoded window-space depth from the WGSL `@builtin(position)` z (the
// VertexOutput.clipPosition lane, which after rasterization holds window-space xyzw with z in NDC depth
// space), linearizes it back into eye space using the camera near/far carried in the material uniform,
// then maps that distance across [near, far] to grayscale `vec3(d)`. NOTE: WebGPU NDC depth is the
// [0, 1] convention (unlike GL's [-1, 1]), so the eye-space reconstruction skips GL's `z * 2 - 1`
// remap — this is the one intentional divergence from depthGlMeshMaterialRenderer's shader.
//
// Normal mode visualizes the WORLD-space surface normal (the geometric normal transformed by the normal
// matrix carried into VertexOutput.worldNormal during vs_main), encoded as `n * 0.5 + 0.5`. NOTE: the
// tangent-space normal map is NOT sampled on wgpu yet — like the StandardPbr / unlit wgpu paths, real
// map textures are not uploaded, so `hasNormalMap` stays false, the bind group binds the shared
// placeholder texture, and the WGSL normal-map branch is reserved for when texture upload lands. This
// mirrors the documented map gap on the rest of the wgpu side.

// The feature flags that select a debug variant. `mode` picks the depth vs normal fragment branch;
// `hasNormalMap` enables the sampled tangent-space normal-map perturbation (normal mode only — depth
// ignores it; not yet wired on wgpu, see the prelude note). Distinct keys compile and cache as distinct
// pipelines.
export interface WgpuDebugDefineKey {
  hasNormalMap: boolean;
  mode: 'depth' | 'normal';
}

// A compiled debug pipeline variant — a WgpuMeshPipeline (pipeline + group(2) material layout).
export interface WgpuDebugPipeline extends WgpuMeshPipeline {}

// Ensures (and caches per material reference) the debug Material bind group — a uniform buffer + the
// shared sampler + the placeholder texture — and rewrites its uniform with this surface's params. The
// params vec4 packs near/far (depth mode) and normalScale (normal mode) into one buffer shared by both
// modes; the active mode reads only the lanes it needs. Mirrors scene-gl's bindGlDebugRange /
// bindGlDebugNormalMap collapsed into one upload. Returns the bind group for the caller to set at
// group(2). normalScale is uploaded for the normal mode; depth mode passes it through unused.
export function bindWgpuDebugSurface(
  state: WgpuRenderState,
  pipeline: Readonly<WgpuDebugPipeline>,
  materialKey: object,
  near: number,
  far: number,
  normalScale: number,
): GPUBindGroup {
  const scene = getWgpuSceneRuntime(state);
  let binding: WgpuMaterialBinding | undefined = scene.materialBindGroups.get(materialKey);
  if (binding === undefined) {
    const stateRuntime = getWgpuRenderStateRuntime(state);
    const buffer = state.device.createBuffer({
      size: DEBUG_UNIFORM_BYTES,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });
    const bindGroup = state.device.createBindGroup({
      layout: pipeline.materialBindGroupLayout,
      entries: [
        { binding: 0, resource: { buffer } },
        { binding: 1, resource: stateRuntime.linearSampler },
        { binding: 2, resource: ensureWgpuPlaceholderTextureView(state) },
      ],
    });
    binding = { bindGroup, buffer };
    scene.materialBindGroups.set(materialKey, binding);
  }

  _scratch[0] = near;
  _scratch[1] = far;
  _scratch[2] = normalScale;
  _scratch[3] = 0;
  state.device.queue.writeBuffer(binding.buffer, 0, _scratch.buffer, 0, DEBUG_UNIFORM_BYTES);
  return binding.bindGroup;
}

// A short, stable, order-independent string identity for a debug define key, used as the pipeline-cache
// key (combined with the color format). Two keys with the same flags produce the same string and so
// share a compiled pipeline. `d-` is depth; `n-` is normal; `nm` is normal + normal map.
export function buildWgpuDebugDefineKey(key: Readonly<WgpuDebugDefineKey>): string {
  return `${key.mode === 'depth' ? 'd' : 'n'}${key.hasNormalMap ? 'm' : '-'}`;
}

// Compiles the debug module for a define key and builds the render pipeline for the given color format,
// with the group(2) material bind-group layout (uniform + sampler + one texture, matching the unlit
// material layout so the shared placeholder satisfies the texture slot). Pure GPU work — no caching —
// used by ensureWgpuDebugPipeline.
export function compileWgpuDebugPipeline(
  state: WgpuRenderState,
  key: Readonly<WgpuDebugDefineKey>,
  format: GPUTextureFormat,
): WgpuDebugPipeline {
  const device = state.device;
  const module = device.createShaderModule({ code: getWgpuDebugModuleSourceForKey(key) });
  const materialBindGroupLayout = device.createBindGroupLayout({
    entries: [
      { binding: 0, visibility: GPUShaderStage.FRAGMENT, buffer: { type: 'uniform' } },
      { binding: 1, visibility: GPUShaderStage.FRAGMENT, sampler: { type: 'filtering' } },
      { binding: 2, visibility: GPUShaderStage.FRAGMENT, texture: { sampleType: 'float' } },
    ],
  });
  // Debug materials are single-sided in OpenFL parity terms; the scene back-face cull is fine for both
  // depth and normal visualization (a culled back face contributes neither depth nor normal).
  return createWgpuMeshPipeline(state, { doubleSided: false, format, materialBindGroupLayout, module });
}

// Resolves the debug pipeline for a define key + color format, compiling and caching it on first use
// through the shared scene pipeline cache under the `debug:` family namespace. Distinct modes (and the
// normal-map variant) cache as distinct entries.
export function ensureWgpuDebugPipeline(
  state: WgpuRenderState,
  key: Readonly<WgpuDebugDefineKey>,
  format: GPUTextureFormat,
): WgpuDebugPipeline {
  return ensureWgpuScenePipeline(state, `debug:${format}|${buildWgpuDebugDefineKey(key)}`, () =>
    compileWgpuDebugPipeline(state, key, format),
  );
}

// The full WGSL module source for a define key: the const-flag block (MODE discriminator + normal-map
// flag) + the shared mesh prelude (Frame/Draw/vs_main/srgbToLinear) + the debug material block +
// fs_main.
export function getWgpuDebugModuleSourceForKey(key: Readonly<WgpuDebugDefineKey>): string {
  return (
    `const MODE : i32 = ${key.mode === 'depth' ? 'DEPTH_MODE' : 'NORMAL_MODE'};\n` +
    `const HAS_NORMAL_MAP : bool = ${key.hasNormalMap ? 'true' : 'false'};\n` +
    DEBUG_MODE_CONSTS_WGSL +
    WGPU_MESH_PRELUDE_WGSL +
    DEBUG_WGSL_BODY
  );
}

// Debug material uniform: params vec4f (16) = 16 bytes / 4 floats. params.x = near, params.y = far
// (depth mode); params.z = normalScale (normal mode, reserved for the future normal-map path).
const DEBUG_UNIFORM_BYTES = 16;

// The two mode discriminator values, declared as consts so the MODE const set by the module header
// reads as a named branch in the fragment body (WGSL has no preprocessor #define).
const DEBUG_MODE_CONSTS_WGSL = /* wgsl */ `
const DEPTH_MODE : i32 = 0;
const NORMAL_MODE : i32 = 1;
`;

const DEBUG_WGSL_BODY = /* wgsl */ `
struct DebugMaterial {
  params : vec4f,  // x = near, y = far (depth); z = normalScale (normal)
};

@group(2) @binding(0) var<uniform> material : DebugMaterial;
@group(2) @binding(1) var materialSampler : sampler;
@group(2) @binding(2) var normalTexture : texture_2d<f32>;

@fragment fn fs_main(in : VertexOutput, @builtin(front_facing) frontFacing : bool) -> @location(0) vec4f {
  if (MODE == DEPTH_MODE) {
    // Linear view-space distance is the perspective w: in.clipPosition is the @builtin(position), whose
    // .w in the fragment stage is 1 / w_clip, so 1 / in.clipPosition.w == w_clip == eye distance. This
    // is camera-agnostic (no camera near/far needed); map it across the material's [near, far]
    // visualization window to grayscale [0, 1].
    let near = material.params.x;
    let far = material.params.y;
    let eyeDepth = 1.0 / in.clipPosition.w;
    let d = clamp((eyeDepth - near) / max(far - near, 1e-6), 0.0, 1.0);
    return vec4f(vec3f(d), 1.0);
  }

  // NORMAL_MODE: visualize the WORLD-space surface normal — the geometric normal carried through
  // draw.normalMatrix in vs_main. The normal-map branch is gated by HAS_NORMAL_MAP but stays inert on
  // wgpu until map upload lands (see the prelude note); normalScale is read so the binding is live.
  var geometricNormal = normalize(in.worldNormal);
  if (!frontFacing) {
    geometricNormal = -geometricNormal;
  }

  var normal = geometricNormal;
  if (HAS_NORMAL_MAP) {
    let tangent = normalize(in.worldTangent.xyz);
    let bitangent = cross(geometricNormal, tangent) * in.worldTangent.w;
    var tangentNormal = textureSample(normalTexture, materialSampler, in.uv).xyz * 2.0 - 1.0;
    tangentNormal = vec3f(tangentNormal.xy * material.params.z, tangentNormal.z);
    let tbn = mat3x3f(tangent, bitangent, geometricNormal);
    normal = normalize(tbn * tangentNormal);
  }

  return vec4f(normal * 0.5 + 0.5, 1.0);
}
`;

const _scratch = new Float32Array(DEBUG_UNIFORM_BYTES / 4);

File: crates/flighthq-scene-wgpu/src/wgpu_debug_prelude.rs
//! The shared Wgpu debug prelude — the WGSL mirror of scene-gl's `glDebugPrelude`.
//! One module source for every lighting-independent debug/utility pass material
//! (Depth, Normal), selected by a `MODE` const discriminator and a `HAS_NORMAL_MAP`
//! const the pipeline compiler folds.
//!
//! Ports `@flighthq/scene-wgpu` `wgpuDebugPrelude.ts`. WebGPU NDC depth is the
//! `[0, 1]` convention, so the depth reconstruction skips GL's `z * 2 - 1` remap —
//! the one intentional divergence from `depthGlMeshMaterialRenderer`'s shader. The
//! tangent-space normal map is not sampled on wgpu yet (`has_normal_map` stays
//! false; the placeholder is bound).

use flighthq_render_wgpu::WgpuRenderState;

use crate::wgpu_mesh_pipeline::{
    CreateWgpuMeshPipelineOptions, WGPU_MESH_PRELUDE_WGSL, WgpuMeshPipeline,
    create_wgpu_mesh_pipeline, ensure_wgpu_placeholder_texture_view,
};
use crate::wgpu_scene_runtime::{WgpuMaterialBinding, WgpuSceneRuntime};

/// A compiled debug pipeline variant.
pub type WgpuDebugPipeline = WgpuMeshPipeline;

/// The debug fragment branch a variant selects. The WGSL mirror of the TS
/// `'depth' | 'normal'` mode string.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum WgpuDebugMode {
    Depth,
    Normal,
}

/// The feature flags that select a debug variant. `mode` picks the depth vs normal
/// fragment branch; `has_normal_map` enables the sampled tangent-space normal-map
/// perturbation (normal mode only; not yet wired on wgpu).
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct WgpuDebugDefineKey {
    pub has_normal_map: bool,
    pub mode: WgpuDebugMode,
}

/// Ensures (once per material kind) the debug Material bind group (a uniform buffer,
/// the shared sampler, and the placeholder texture) and rewrites its uniform. The
/// params vec4 packs near/far (depth mode) and normalScale (normal mode); the active
/// mode reads only the lanes it needs. Mirrors TS `bindWgpuDebugSurface` (keyed by
/// kind + threaded runtime + `cache_key` layout lookup).
pub fn bind_wgpu_debug_surface(
    state: &WgpuRenderState,
    scene: &mut WgpuSceneRuntime,
    cache_key: &str,
    material_key: flighthq_types::kind::KindId,
    near: f32,
    far: f32,
    normal_scale: f32,
) {
    if !scene.material_bind_groups.contains_key(&material_key) {
        let buffer = state.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("flight-wgpu-debug-material-uniform"),
            size: DEBUG_UNIFORM_BYTES,
            usage: wgpu::BufferUsages::UNIFORM | wgpu::BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });
        ensure_wgpu_placeholder_texture_view(state, scene);
        let bind_group = {
            let layout = &scene.mesh_pipeline_cache[cache_key].material_bind_group_layout;
            let placeholder = scene
                .placeholder_view
                .as_ref()
                .expect("placeholder view ensured before material bind group");
            state.device.create_bind_group(&wgpu::BindGroupDescriptor {
                label: Some("flight-wgpu-debug-material-bind-group"),
                layout,
                entries: &[
                    wgpu::BindGroupEntry {
                        binding: 0,
                        resource: buffer.as_entire_binding(),
                    },
                    wgpu::BindGroupEntry {
                        binding: 1,
                        resource: wgpu::BindingResource::Sampler(&state.runtime.linear_sampler),
                    },
                    wgpu::BindGroupEntry {
                        binding: 2,
                        resource: wgpu::BindingResource::TextureView(placeholder),
                    },
                ],
            })
        };
        scene
            .material_bind_groups
            .insert(material_key, WgpuMaterialBinding { bind_group, buffer });
    }

    let binding = &scene.material_bind_groups[&material_key];
    let scratch = [near, far, normal_scale, 0.0];
    state
        .queue
        .write_buffer(&binding.buffer, 0, f32_slice_bytes(&scratch));
}

/// A short, stable, order-independent string identity for a debug define key. `d-`
/// is depth; `n-` is normal; `nm` is normal + normal map.
pub fn build_wgpu_debug_define_key(key: &WgpuDebugDefineKey) -> String {
    format!(
        "{}{}",
        if key.mode == WgpuDebugMode::Depth {
            'd'
        } else {
            'n'
        },
        if key.has_normal_map { 'm' } else { '-' },
    )
}

/// Compiles the debug module for a define key and builds the render pipeline for the
/// color format, with the group(2) material bind-group layout (uniform + sampler +
/// one texture). Debug materials are single-sided. Mirrors TS
/// `compileWgpuDebugPipeline`.
pub fn compile_wgpu_debug_pipeline(
    state: &mut WgpuRenderState,
    scene: &mut WgpuSceneRuntime,
    key: &WgpuDebugDefineKey,
    format: wgpu::TextureFormat,
) -> WgpuDebugPipeline {
    let device = &state.device;
    let module = device.create_shader_module(wgpu::ShaderModuleDescriptor {
        label: Some("flight-wgpu-debug-shader"),
        source: wgpu::ShaderSource::Wgsl(get_wgpu_debug_module_source_for_key(key).into()),
    });
    let material_bind_group_layout =
        device.create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
            label: Some("flight-wgpu-debug-material-bgl"),
            entries: &[
                wgpu::BindGroupLayoutEntry {
                    binding: 0,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Buffer {
                        ty: wgpu::BufferBindingType::Uniform,
                        has_dynamic_offset: false,
                        min_binding_size: None,
                    },
                    count: None,
                },
                wgpu::BindGroupLayoutEntry {
                    binding: 1,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Sampler(wgpu::SamplerBindingType::Filtering),
                    count: None,
                },
                wgpu::BindGroupLayoutEntry {
                    binding: 2,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Texture {
                        sample_type: wgpu::TextureSampleType::Float { filterable: true },
                        view_dimension: wgpu::TextureViewDimension::D2,
                        multisampled: false,
                    },
                    count: None,
                },
            ],
        });
    create_wgpu_mesh_pipeline(
        state,
        scene,
        CreateWgpuMeshPipelineOptions {
            double_sided: false,
            format,
            material_bind_group_layout,
            module,
            topology: wgpu::PrimitiveTopology::TriangleList,
        },
    )
}

/// Resolves the debug pipeline for a define key + color format, compiling and
/// caching it on first use, and returns its `mesh_pipeline_cache` key. Mirrors TS
/// `ensureWgpuDebugPipeline` (returns the string key — see
/// `ensure_wgpu_classic_pipeline`).
pub fn ensure_wgpu_debug_pipeline(
    state: &mut WgpuRenderState,
    scene: &mut WgpuSceneRuntime,
    key: &WgpuDebugDefineKey,
    format: wgpu::TextureFormat,
) -> String {
    let cache_key = format!("debug:{format:?}|{}", build_wgpu_debug_define_key(key));
    if !scene.mesh_pipeline_cache.contains_key(&cache_key) {
        let pipeline = compile_wgpu_debug_pipeline(state, scene, key, format);
        scene
            .mesh_pipeline_cache
            .insert(cache_key.clone(), pipeline);
    }
    cache_key
}

/// The full WGSL module source for a define key.
pub fn get_wgpu_debug_module_source_for_key(key: &WgpuDebugDefineKey) -> String {
    format!(
        "const MODE : i32 = {};\n\
         const HAS_NORMAL_MAP : bool = {};\n\
         {}{}{}",
        if key.mode == WgpuDebugMode::Depth {
            "DEPTH_MODE"
        } else {
            "NORMAL_MODE"
        },
        bool_literal(key.has_normal_map),
        DEBUG_MODE_CONSTS_WGSL,
        WGPU_MESH_PRELUDE_WGSL,
        DEBUG_WGSL_BODY,
    )
}

const DEBUG_UNIFORM_BYTES: u64 = 16;

const DEBUG_MODE_CONSTS_WGSL: &str = r#"
const DEPTH_MODE : i32 = 0;
const NORMAL_MODE : i32 = 1;
"#;

const DEBUG_WGSL_BODY: &str = r#"
struct DebugMaterial {
  params : vec4f,  // x = near, y = far (depth); z = normalScale (normal)
};

@group(2) @binding(0) var<uniform> material : DebugMaterial;
@group(2) @binding(1) var materialSampler : sampler;
@group(2) @binding(2) var normalTexture : texture_2d<f32>;

@fragment fn fs_main(in : VertexOutput, @builtin(front_facing) frontFacing : bool) -> @location(0) vec4f {
  if (MODE == DEPTH_MODE) {
    // Linear view-space distance is the perspective w: in.clipPosition is the @builtin(position), whose
    // .w in the fragment stage is 1 / w_clip, so 1 / in.clipPosition.w == w_clip == eye distance. This
    // is camera-agnostic (no camera near/far needed); map it across the material's [near, far]
    // visualization window to grayscale [0, 1].
    let near = material.params.x;
    let far = material.params.y;
    let eyeDepth = 1.0 / in.clipPosition.w;
    let d = clamp((eyeDepth - near) / max(far - near, 1e-6), 0.0, 1.0);
    return vec4f(vec3f(d), 1.0);
  }

  // NORMAL_MODE: visualize the WORLD-space surface normal — the geometric normal carried through
  // draw.normalMatrix in vs_main. The normal-map branch is gated by HAS_NORMAL_MAP but stays inert on
  // wgpu until map upload lands (see the prelude note); normalScale is read so the binding is live.
  var geometricNormal = normalize(in.worldNormal);
  if (!frontFacing) {
    geometricNormal = -geometricNormal;
  }

  var normal = geometricNormal;
  if (HAS_NORMAL_MAP) {
    let tangent = normalize(in.worldTangent.xyz);
    let bitangent = cross(geometricNormal, tangent) * in.worldTangent.w;
    var tangentNormal = textureSample(normalTexture, materialSampler, in.uv).xyz * 2.0 - 1.0;
    tangentNormal = vec3f(tangentNormal.xy * material.params.z, tangentNormal.z);
    let tbn = mat3x3f(tangent, bitangent, geometricNormal);
    normal = normalize(tbn * tangentNormal);
  }

  return vec4f(normal * 0.5 + 0.5, 1.0);
}
"#;

fn bool_literal(value: bool) -> &'static str {
    if value { "true" } else { "false" }
}

fn f32_slice_bytes(data: &[f32]) -> &[u8] {
    let ptr = data.as_ptr() as *const u8;
    unsafe { std::slice::from_raw_parts(ptr, std::mem::size_of_val(data)) }
}

#[cfg(test)]
mod tests {
    use super::*;

    mod build_wgpu_debug_define_key {
        use super::*;

        #[test]
        fn encodes_mode_then_normal_map_flag() {
            assert_eq!(
                build_wgpu_debug_define_key(&WgpuDebugDefineKey {
                    has_normal_map: false,
                    mode: WgpuDebugMode::Depth,
                }),
                "d-"
            );
            assert_eq!(
                build_wgpu_debug_define_key(&WgpuDebugDefineKey {
                    has_normal_map: true,
                    mode: WgpuDebugMode::Normal,
                }),
                "nm"
            );
        }
    }

    mod get_wgpu_debug_module_source_for_key {
        use super::*;

        #[test]
        fn selects_the_depth_mode_discriminator_and_carries_both_branches() {
            let source = get_wgpu_debug_module_source_for_key(&WgpuDebugDefineKey {
                has_normal_map: false,
                mode: WgpuDebugMode::Depth,
            });
            assert!(source.contains("const MODE : i32 = DEPTH_MODE;"));
            assert!(source.contains("const DEPTH_MODE : i32 = 0;"));
            assert!(source.contains("let eyeDepth = 1.0 / in.clipPosition.w;"));
            assert!(source.contains("return vec4f(normal * 0.5 + 0.5, 1.0);"));
        }

        #[test]
        fn selects_the_normal_mode_discriminator() {
            let source = get_wgpu_debug_module_source_for_key(&WgpuDebugDefineKey {
                has_normal_map: true,
                mode: WgpuDebugMode::Normal,
            });
            assert!(source.contains("const MODE : i32 = NORMAL_MODE;"));
            assert!(source.contains("const HAS_NORMAL_MAP : bool = true;"));
        }
    }
}
